### DISCUSSION TITLE
Webgl 2.0

### DISCUSSION THREAD

**Discussion by MasonJohnHawver42** [raysan5/raylib#2447] [Ideas]
Will Raylib ever use Webgl 2.0? is there a plan to use it? is there a reason why we still are using Webgl 1.0?

**Comment by raysan5** [comment]
raylib only supports OpenGL ES 2.0 that maps to WebGL 1.0, that is probably more than enough for most raylib users. What would be the benefits of supporting WebGL 2.0 vs the cost of implementing it?

**Comment by MasonJohnHawver42** [comment]
Where can I start implementing it? 
Edit: The same solution can be used that supports Opengl 4.3 because Webgl 2 is backwards compatible, right?

**Reply by MasonJohnHawver42** [reply]
Note: Your right there is no explicit feature that Webgl 2 offers that I need. I just wan't to use glsl 330 because there are tutorials for it.

**Reply by bosoni** [reply]
@MasonJohnHawver42   you can write transpiler which converts gles3-> gles2 (did years ago gl2->gl3). Just found this link https://stackoverflow.com/questions/42231698/how-to-convert-glsl-version-330-core-to-glsl-es-version-100


"Webgl 2 is backwards compatible, right?"
   https://www.khronos.org/registry/webgl/specs/latest/2.0/#1
      ""WebGL 2.0 is not entirely backwards compatible with WebGL 1.0""


WEBGL 2.0:  MRT (multi render targets) comes to mind (if using webgl2.0) but yeah,  I dont think MRT isn'n needed most users (if any), and it is easier to create multiple textures and using couple of shaders if need to make deferred renderer (occlusion culling etc so render fast gbuffer (when only visible objects are rendered) and then shader which uses all textures and outputs final image). 
(Personally, I dont need webgl2, just babling around :D )

**Reply by KTRosenberg** [reply]
There are many webgl2 features that would be great to use, actually.

**Reply by raysan5** [reply]
@KTRosenberg I think there are not that many features most raylib users need.

**Reply by chemaguerra** [reply]
Many mobile phones don't support FP32 texture extensions (despite they often support WebGL2, which does without any extensions). So support for floating-point textures to do HDR, compute shading, etc... would be my main reason to strongly wish for WebGL2 support. MRTs would be a close second.

**Reply by raysan5** [reply]
Just added some flags to support OpenGL ES 3.0 context creation (despite limited to OpenGL ES 2.0 functionality for now). It has not been tested, feel free to report any issue: https://github.com/raysan5/raylib/commit/e497603678823dc93f01dd9686819a92d780ad67

**Reply by chemaguerra** [reply]
You deserve a truckload of gratitude. :)

I did indeed test all this and have made some good progress. Got my code running on WebGL2 + ES 300 + floating-point textures + Multiple-Render-Targets. I had to complete some stuff related to the initialization flags, though. I will follow up on https://github.com/raysan5/raylib/commit/e497603678823dc93f01dd9686819a92d780ad67 shortly.

**Reply by raysan5** [reply]
@chemaguerra Good to know that it somewhat worked, feel free to send a PR if some part require to be reviewed. Additional functionality could be exposed just using the `#if defined(GRAPHICS_API_OPENGL_ES3)`.

**Reply by chemaguerra** [reply]
I finally gathered some time to clean-up and test my changes properly. I sent a PR: https://github.com/raysan5/raylib/pull/3107

**Reply by raysan5** [reply]
@chemaguerra Thank you very much! Just added a couple of comments to review!

**Comment by chemaguerra** [comment]
I made some amends following your comments.

**Reply by raysan5** [reply]
Thanks for the review! PR merged!
